### DISCUSSION TITLE
Row level security disallowing insert?

### DISCUSSION THREAD

**Discussion by prescience-data** [supabase/supabase#3023] [Questions] [answered]
I'm having an issue when attempting to insert a new row given the following schema.

For quick context, there are three tables (two actual and one pivot):

- `public.accounts`
- `public.memberships`
- `public.teams`

The RLS is set to only allow `SELECT` from `teams` provided the currently authenticated user (`account`) has a `membership` pivot connecting them to that `team`.

From what [I read here](https://nikofischer.com/supabase-error-row-level-security-policy) it seems as though the `SELECT` policy is kicking in during the `INSERT` (as it calls `RETURNING INTO` to get the newly inserted `teams.id` to then create the `membership`. 

Unfortunately, this creates a circular loop as the `teams` `SELECT` policy requires a `memberships` entry, but the `memberships` entry cannot be created without the `teams.id`.

😵‍💫😵

Hopefully, this is just a "you're doing it wrong" problem and there is a more logical way to do this...

```pgsql
-- CreateFunction
CREATE OR REPLACE FUNCTION create_team
(
  name text,
  slug text,
  country text,
  is_default boolean
)
  RETURNS team_filter
  LANGUAGE plpgsql
AS
$$
  # VARIABLE_CONFLICT USE_COLUMN
DECLARE
  _team_id       bigint;
  _membership_id bigint;
BEGIN

  INSERT INTO teams ( name, slug, country ) VALUES ( $1::text, $2::text, $3::text ) RETURNING teams.id INTO _team_id;  
  IF _team_id IS NULL THEN
    RAISE EXCEPTION 'Team failed to insert.';
  END IF;
  
  INSERT INTO memberships ( account_id, team_id, is_default ) VALUES ( auth.uid( ), _team_id, $4::boolean ) RETURNING memberships.id INTO _membership_id;
  IF _membership_id IS NULL THEN
    RAISE EXCEPTION 'Failed to create membership.';
  END IF;

  RETURN fetch_team( _team_id );

END;
$$;

--------------------------------------------------------------------------------

-- CreateFunction
CREATE OR REPLACE FUNCTION __check__account_has_team
(
  account_id uuid,
  team_id bigint
)
  RETURNS boolean
  SECURITY DEFINER
  LANGUAGE plpgsql
AS
$$
DECLARE
  _has_membership boolean;
BEGIN
  SELECT COUNT( memberships.id ) > 0
  INTO _has_membership
  FROM memberships
  WHERE memberships.account_id = $1::uuid
    AND memberships.team_id = $2::bigint;

  RETURN _has_membership;
END;
$$ STABLE;

--------------------------------------------------------------------------------

-- EnableSecurity
ALTER TABLE teams
  ENABLE ROW LEVEL SECURITY;

-- CreatePolicy
CREATE POLICY "authenticated accounts can view their own account teams. " ON teams FOR
  SELECT USING (__check__account_has_team( auth.uid( ), teams.id ));

-- CreatePolicy
CREATE POLICY "authenticated accounts can update their own account teams." ON teams FOR
  UPDATE USING (__check__account_has_team( auth.uid( ), teams.id ));

-- CreatePolicy
CREATE POLICY "authenticated accounts can delete their own teams." ON teams FOR
  DELETE USING (__check__account_has_team( auth.uid( ), teams.id ));

```

**Comment by steve-chavez** [comment]
> From what I read here it seems as though the SELECT policy is kicking in during the INSERT (as it calls RETURNING INTO to get the newly inserted teams.id to then create the membership.

Wouldn't this be solved with `return=minimal`? Then SELECT won't kick in with INSERT.

```js
await supabase.from('sample').insert({ key: 'value' }, { returning: 'minimal' })
```

**Reply by prescience-data** [reply]
It's a function, not a direct query, unfortunately.  I'm attempting to build the API in a way where only functions are exposed (unless this is super anti-pattern?)

Additionally, I'm not sure how that would help, as I still need the team id to insert the membership, so it needs to be `SELECT`ed  from at some point (which in turn causes the recursive logic).

eg cannot insert the `membership` without the `team.id`, and cannot get the `team.id` without the `membership` being already created.

**Reply by w3b6x9** [reply]
@prescience-data you can update your functions to be `SECURITY DEFINER` ones, and if you did it in the Supabase dashboard's SQL editor, then no matter who calls your functions they'll have permission to do what you want. Just make sure to reference `Writing SECURITY DEFINER Functions Safely` section of https://www.postgresql.org/docs/13/sql-createfunction.html for best practices.

**Reply by prescience-data** [reply]
What I am currently testing is a private `SECURITY DEFINER` function that has the same param signature + a param for `account_id`, which is not directly exposed. 

Then the public exposed `SECURITY INVOKER` function just passes all the desired params, along with the `account_id` produced by `auth.uid()`.

Fingers crossed! 🤞🏻

**Reply by steve-chavez** [reply]
>  I'm attempting to build the API in a way where only functions are exposed (unless this is super anti-pattern?)

Just want to point out that only exposing functions is a valid model, check this previous discussion https://github.com/supabase/supabase/discussions/1360

**Comment by prescience-data** [comment, marked as answer]
> @prescience-data you can update your functions to be `SECURITY DEFINER` ones, and if you did it in the Supabase dashboard's SQL editor, then no matter who calls your functions they'll have permission to do what you want. Just make sure to reference `Writing SECURITY DEFINER Functions Safely` section of https://www.postgresql.org/docs/13/sql-createfunction.html for best practices.


For anyone searching this in the future, this was the solution thanks to @w3b6x9 .

```pgsql

CREATE OR REPLACE FUNCTION __private__create_team_for_account
(
  account_id uuid,
  name text,
  slug text,
  country text
)
  RETURNS bigint
  SECURITY DEFINER
  LANGUAGE plpgsql
AS
$$
  # VARIABLE_CONFLICT USE_COLUMN
DECLARE
  _team_id       bigint;
  _membership_id bigint;
BEGIN
  INSERT INTO teams ( name, slug, country )
  VALUES ( $2::text, $3::text, $4::text )
  RETURNING id INTO _team_id;
  IF _team_id IS NULL THEN
    RAISE EXCEPTION 'Failed to create team for account %.', $1::text;
  END IF;

  INSERT INTO memberships ( account_id, team_id )
  VALUES ( $1::uuid, _team_id::bigint )
  RETURNING id INTO _membership_id;
  IF _membership_id IS NULL THEN
    RAISE EXCEPTION 'Failed to create membership for team %.', _team_id;
  END IF;

  RETURN _team_id;
END;
$$;


-- CreateFunction
CREATE OR REPLACE FUNCTION create_team
(
  name text,
  slug text,
  country text,
  is_default boolean
)
  RETURNS team_filter
  LANGUAGE plpgsql
AS
$$
  # VARIABLE_CONFLICT USE_COLUMN
DECLARE
  _team_id              bigint;
  _has_set_default_team boolean;
BEGIN

  SELECT __private__create_team_for_account( auth.uid( ), $1::text, $2::text, $3::text ) INTO _team_id;

  IF $4::boolean IS TRUE THEN
    SELECT set_default_team( _team_id ) INTO _has_set_default_team;
  END IF;

  RETURN fetch_team( _team_id );

END;
$$;
```

**Reply by w3b6x9** [reply]
nice, glad you got it working!
